sys/xtimer/xtimer_core.c

```diff
@@ -287,31 +188,46 @@ static int _remove_timer_from_list(xtimer_t **list_head, xtimer_t *timer)
     return 0;
 }
 
+/**
+ * @brief remove a timer only when it is not about to expire
+ *        this function is time critical and must be interrupt free
+ */
 static void _remove(xtimer_t *timer)
 {
-    if (timer_list_head == timer) {
-        uint32_t next;
+    if (timer == timer_list_head) {
         timer_list_head = timer->next;
+        timer->offset = 0;
+        timer->long_offset = 0;
+        timer->next = NULL;
+
+        uint32_t now = _xtimer_now();
+        _update_short_timers(&now);
+
         if (timer_list_head) {
-            /* schedule callback on next timer target time */
-            next = timer_list_head->target - XTIMER_OVERHEAD;
+            if (timer_list_head->offset <= _xtimer_lltimer_mask(0xFFFFFFFF)) {
+                /* schedule callback on next timer target time */
+                _lltimer_set(timer_list_head->start_time + timer_list_head->offset);
+            }
+            else if (!_lltimer_ongoing) {
+                /* schedule callback after max_low_level_time/2 to detect a cycle */
+                _lltimer_set(_xtimer_now() + (_xtimer_lltimer_mask(0xFFFFFFFF)>>1));
+            }
         }
-        else {
-            next = _xtimer_lltimer_mask(0xFFFFFFFF);
+        else if (!_lltimer_ongoing) {
+            /* schedule callback after max_low_level_time/2 to detect a cycle */
+            _lltimer_set(_xtimer_now() + (_xtimer_lltimer_mask(0xFFFFFFFF)>>1));
         }
-        _lltimer_set(next);
     }
     else {
         if (!_remove_timer_from_list(&timer_list_head, timer)) {
-            if (!_remove_timer_from_list(&overflow_list_head, timer)) {
-                _remove_timer_from_list(&long_list_head, timer);
-            }
+            _remove_timer_from_list(&long_list_head, timer);
         }
     }
 }
 
 void xtimer_remove(xtimer_t *timer)
 {
+    /* time sensitive since the target timer can be fired */
     int state = irq_disable();
```

Comment:
just to make it consistent this should also be `unsigned int`

Reply:
done